src/v/cluster/distributed_kv_stm.h

```diff
@@ -311,6 +312,17 @@ class distributed_kv_stm final : public raft::persisted_stm<> {
             // the write lock we hold.
             co_await ss::yield();
         }
+        if (deleted.size() == 0) {
+            co_return errc::success;
+        }
+        // downgrade the lock to a read lock so replicate_and_wait can make
+        // progress.
+        units.return_all();
+        auto read_units = _snapshot_lock.attempt_read_lock();
+        vassert(
+          read_units,
```

Comment:
i was under the impression that this wouldn't work because the underlying semaphore attempts to prevent starvation by forcing new acquirers to queue up behind other waiters?

Reply:
ya you're right I think (nice catch), a sequence of events where this could fail.

u = write_lock (current),  queue = [write_lock, ...] all the units deposited by current are grabbed by the first waiter.

let me fix this.